config/config_params_test.go

```diff
@@ -427,6 +431,70 @@ var _ = Describe("DatastoreConfig tests", func() {
 			Expect(c.DatastoreConfig().Spec.K8sDisableNodePoll).To(BeTrue())
 		})
 	})
+	Describe("with the configuration set only from the common calico configuration", func() {
+		BeforeEach(func() {
+			os.Setenv("CALICO_DATASTORE_TYPE", "etcdv3")
+			os.Setenv("CALICO_ETCD_ENDPOINTS", "http://localhost:1234")
+			os.Setenv("CALICO_ETCD_KEY_FILE", testutils.TestDataFile("etcdkeyfile.key"))
+			os.Setenv("CALICO_ETCD_CERT_FILE", testutils.TestDataFile("etcdcertfile.cert"))
+			os.Setenv("CALICO_ETCD_CA_CERT_FILE", testutils.TestDataFile("etcdcacertfile.cert"))
```

Comment:
I'm wondering why these `os.Setenv` calls work with ginkgo executing tests in an arbitrary order, and possibly in multiple parallel processes.  The concern is that they change global state and don't reset it for other test cases.

Reply:
I see the tests are written so as not to be sensitive to this.  All the same, if there's something better we can easily do, let's do that; or at least comment about this.

Reply:
I think we need to use the environment here, if you look at line 401 in config_params.go it loads the calico configuration from the environment, so this test is making sure that it actually loads them from the environment and that the DataStoreConfig returned is filled with these values. I understand the concern, and will add a comment about it, and I'm definitely open to other suggestions about how to test this.

Reply:
General point: this kind of thing comes up a lot in unit tests.  Remember that this is unit testing, and generally, with unit testing, what we're really trying to test is only the code directly in front of us in a given unit (= .go file or package).  When that code has external inputs, like `apiconfig.LoadClientConfigFromEnvironment()`, we can often control those inputs for test purposes by insinuating a shim or mock at that point.

In this specific case, what we could do is:
- add a `getConfigFromEnvironment func() (*apiconfig.CalicoAPIConfig, error)` field in `Config`
- for production, set `getConfigFromEnvironment` to `apiconfig.LoadClientConfigFromEnvironment`
- for unit testing, have a way for the UT code to provide an alternative function implementation, under the UT code's control.

That's the kind of thing that we often do.  The question is just whether it's worth it in this case; I think I'm leaning towards a yes.  WDYT?